packages/labs/signals/src/lib/signal-watcher.ts

```diff
@@ -48,24 +101,61 @@ const elementForWatcher = new WeakMap<
  * watch for access to signals during the update lifecycle and trigger a new
  * update when signals values change.
  */
-export function SignalWatcher<T extends ReactiveElementConstructor>(
-  Base: T
-): T {
+export function SignalWatcher<T extends Constructor<ReactiveElement>>(Base: T) {
   // Only apply the mixin once
   if ((Base as typeof SignalWatcher)[signalWatcherBrand] === true) {
     console.warn(
       'SignalWatcher should not be applied to the same class more than once.'
     );
-    return Base;
+    return Base as T & Constructor<SignalWatcherApi>;
   }
 
-  abstract class SignalWatcher extends Base implements SignalWatcherInterface {
+  class SignalWatcher extends Base implements SignalWatcherInterface {
     static [signalWatcherBrand]: true;
 
-    private __watcher?: Signal.subtle.Watcher;
+    // @internal used in watch directive
+    _watcher?: Signal.subtle.Watcher;
+
+    /**
+     * Flushes effects in required order:
+     * 1. Before update effects
+     * 2. Perform update
+     * 3. Pending watches
+     * 4. After update effects
+     * */
+    private __flushEffects() {
+      const beforeEffects = [] as Signal.Computed<void>[];
+      const afterEffects = [] as Signal.Computed<void>[];
+      this.__effects.forEach((options, signal) => {
+        const list = options?.beforeUpdate ? beforeEffects : afterEffects;
+        list.push(signal);
+      });
+      const pendingWatches = this._watcher
+        ?.getPending()
+        .filter(
+          (signal) =>
+            signal !== this.__performUpdateSignal && !this.__effects.has(signal)
+        );
+      beforeEffects.forEach((signal) => signal.get());
+      this.__performUpdateSignal?.get();
+      pendingWatches!.forEach((signal) => signal.get());
```

Comment:
Should the pending watches only be run if the full update didn't run? Otherwise those bindings will run twice?

Reply:
No, the get won't provoke it running twice if the value hasn't changed, right?

Reply:
ah, correct